File: catkin_ws/src/sailrobot/src/utilities.cpp
#include <utilities.hpp>
#include <math.h>
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <iterator>
#include <algorithm>

using namespace glm;

float Utility::GPSDist(float lat1, float lon1, float lat2, float lon2){
    float R = 6371000;
    float ksi1 = lat1*M_PI/180.0;
    float ksi2 = lat2*M_PI/180.0;
    float dksi = (lat2-lat1)*M_PI/180.0;
    float dlambda = (lon2-lon1)*M_PI/180.0;
    
    float a = sin(dksi/2.0)*sin(dksi/2.0) + cos(ksi1)*cos(ksi2)*sin(dlambda/2.0)*sin(dlambda/2.0);
    float c = 2*atan2(sqrt(a), sqrt(1-a));
    return R*c;
}

float Utility::GPSDist(vec2 point1, vec2 point2){
    float R = 6371000;
    float ksi1 = point1.x*M_PI/180.0;
    float ksi2 = point2.x*M_PI/180.0;
    float dksi = (point2.x-point1.x)*M_PI/180.0;
    float dlambda = (point2.y-point1.y)*M_PI/180.0;

    float a = sin(dksi/2.0)*sin(dksi/2.0) + cos(ksi1)*cos(ksi2)*sin(dlambda/2.0)*sin(dlambda/2.0);
    float c = 2*atan2(sqrt(a), sqrt(1-a));
    return R*c;
}

float Utility::GPSDistFast(float lat1, float lon1, float lat2, float lon2){
    float ksi1 = lat1*M_PI/180.0;
    float ksi2 = lat2*M_PI/180.0;
    float lam1 = lon1*M_PI/180.0;
    float lam2 = lon2*M_PI/180.0;

    float x = (lam2-lam1)*cos((ksi1+ksi2)/2.0);
    float y = ksi2-ksi1;
    return sqrt(x*x + y*y) * 6371000;
}

float Utility::GPSDistFast(vec2 point1, vec2 point2){
    float ksi1 = point1.x*M_PI/180.0;
    float ksi2 = point2.x*M_PI/180.0;
    float lam1 = point1.y*M_PI/180.0;
    float lam2 = point2.y*M_PI/180.0;
    
    float x = (lam2-lam1)*cos((ksi1+ksi2)/2.0);
    float y = ksi2-ksi1;
    return sqrt(x*x + y*y) * 6371000;
}

float Utility::GPSBearing(float lat1, float lon1, float lat2, float lon2){
    float ksi1 = lat1*M_PI/180.0;
    float ksi2 = lat2*M_PI/180.0;
    float lam1 = lon1*M_PI/180.0;
    float lam2 = lon2*M_PI/180.0;
    
    float y = sin(lam2-lam1)*cos(ksi2);
    float x = cos(ksi1)*sin(ksi2)-sin(ksi1)*cos(ksi2)*cos(lam2-lam1);
    return -atan2(y,x);
}

float Utility::GPSBearing(vec2 point1, vec2 point2){
    float ksi1 = point1.x*M_PI/180.0;
    float ksi2 = point2.x*M_PI/180.0;
    float lam1 = point1.y*M_PI/180.0;
    float lam2 = point2.y*M_PI/180.0;

    float y = sin(lam2-lam1)*cos(ksi2);
    float x = cos(ksi1)*sin(ksi2)-sin(ksi1)*cos(ksi2)*cos(lam2-lam1);
    return -atan2(y,x);
}

float Utility::GPStoCartesian(float lat, float gpslong){
    
}

float Utility::CartesiantoGPS(float x, float y){
    
}

vec3 Utility::QuaternionToEuler(float x, float y, float z, float w){
    return eulerAngles(quat(x,y,z,w));
}

quat Utility::EulerToQuaternion(float x, float y, float z){
    return quat(vec3(x,y,z));
}

vec2* Utility::ReadGPSCoordinates(std::string filepath){
    vec2* coordinates = NULL;
    std::fstream file(filepath);
    if(!file){
        std::cerr << "GPS Reading : Couldn't open file" << std::endl;
        return coordinates;
    }
    std::string line;
    int nbLines = std::count(std::istreambuf_iterator<char>(file),
                             std::istreambuf_iterator<char>(), '\n');
    coordinates = new vec2[nbLines];
    file.clear();
    file.seekg(0, std::ios::beg);
    int i = 0;
    while (std::getline(file, line)) {
        std::string coords;
        std::stringstream stream(line);
        std::getline(stream, coords, ',');
        coordinates[i].x = std::atof(coords.c_str());
        std::getline(stream, coords, ',');
        coordinates[i].y = std::atof(coords.c_str());
        i++;
    }
    file.close();
    return coordinates;
}

File: catkin_ws/src/sailrobot/scripts/utilities.py
#!/usr/bin/env python
import math
import numpy as np
import os
import rospkg

def GPSDist(lat1, lon1, lat2, lon2):
        R = 6371000
        degtorad = math.pi/180.0
        ksi1 = lat1*degtorad
        ksi2 = lat2*degtorad
        dksi = (lat2-lat1)*degtorad
        dlambda = (lon2-lon1)*degtorad

        a = math.sin(dksi/2.0)*math.sin(dksi/2.0) + math.cos(ksi1)*math.cos(ksi2)*math.sin(dlambda/2.0)*math.sin(dlambda/2.0)
        c = 2*math.atan2(math.sqrt(a), math.sqrt(1-a))
        return R*c

def GPSDistFast(lat1, lon1, lat2, lon2):
        degtorad = math.pi/180.0
        ksi1 = lat1*degtorad
        ksi2 = lat2*degtorad
        lam1 = lon1*degtorad
        lam2 = lon2*degtorad

        x = (lam2-lam1)*math.cos((ksi1+ksi2)/2.0)
        y = ksi2-ksi1
        return math.sqrt(x*x + y*y) * 6371000

def GPSBearing(lat1, lon1, lat2, lon2):
        degtorad = math.pi/180.0
        ksi1 = lat1*degtorad
        ksi2 = lat2*degtorad
        lam1 = lon1*degtorad
        lam2 = lon2*degtorad

        y = math.sin(lam2-lam1)*math.cos(ksi2)
        x = math.cos(ksi1)*math.sin(ksi2)-math.sin(ksi1)*math.cos(ksi2)*math.cos(lam2-lam1)
	bearing = -math.atan2(y,x)
        return bearing

def QuaternionToEuler(x,y,z,w):
        zsqr = z * z
	
        t0 = +2.0 * (x * y + z * w)
        t1 = +1.0 - 2.0 * (y * y + zsqr)
        X = math.atan2(t0, t1)
	
        t2 = +2.0 * (x * z - w * y)
        t2 = +1.0 if t2 > +1.0 else t2
        t2 = -1.0 if t2 < -1.0 else t2
        Y = math.asin(t2)
	
        t3 = +2.0 * (x * w + y * z)
        t4 = +1.0 - 2.0 * (zsqr + w * w)
        Z = math.atan2(t3, t4)
	
        return X, Y, Z

def EulerToQuaternion(x, y, z):
        cy = math.cos(z * 0.5)
        sy = math.sin(z * 0.5)
        cr = math.cos(y * 0.5)
        sr = math.sin(y * 0.5)
        cp = math.cos(x * 0.5)
        sp = math.sin(x * 0.5)

        qx = cy * cr * cp + sy * sr * sp
        qy = cy * sr * cp - sy * cr * sp
        qz = cy * cr * sp + sy * sr * cp
        qw = sy * cr * cp - cy * sr * sp
	
        return qx,qy,wz,qw

def readGPSCoordinates(filepath):
	rospack = rospkg.RosPack()
	rospath = rospack.get_path('sailrobot')
	if filepath[0] is not '/':
		filepath = rospath + "/" + filepath
        if os.path.exists(filepath):
                with open(filepath, 'r') as file:
                        try:
                                res = []
                                noproblem = True
                                for line in file:
                                        coords = line.split(",")
                                        try:
                                                res.append([float(coords[0]),float(coords[1])])
                                        except:
                                                noproblem = False
                                if not noproblem:
                                        print("Reading GPS : Wrong File Format for some lines")
                                return res
                        except:
                                print("Reading GPS : File not opened")
        else:
                print("Reading GPS : File does not exist")
        return []
